docs/shared-constraints.md

```diff
@@ -0,0 +1,69 @@
+# Shared Constraints
+
+Custom constraints in `protovalidate` afford a lot of power, but can become
+cumbersome and repetitive when the same kind of custom constraints are needed
+across multiple fields or messages. To this end, `protovalidate` provides a
+mechanism for creating reusable constraints that can be applied on multiple
+fields.
+
+Shared constraints require Protobuf extensions, which are not available in
+proto3. Either proto2 or at least Protobuf 2023 Edition must be used to define
+shared constraints. Shared constraints defined in proto2 or Protobuf 2023
+Edition or later can be imported and utilized in proto3 files.
+
+## Shared Field Constraints
+
+To define a shared field rule, extend one of the standard rules messages. For
+example, to define a new rule for `float` fields, extend
+`buf.validate.FloatRules`, as follows:
+
+```proto
+import "buf/validate/shared/constraints.proto";
+import "buf/validate/validate.proto";
+
+extend buf.validate.FloatRules {
+  float abs_range = 80048952 [(buf.validate.shared.field).cel = {
+    id: "float.abs_range"
+    expression: "this >= -rule && this <= rule"
+    message: "float value is out of range"
+  }];
+}
+```
+
+> [!TIP]
+> Rules can refer to their own value with the `rule` constant.
+
+> [!WARNING]
+> Be mindful that extension numbers must not conflict with any other extension
+> to the same message across _all_ Protobuf files in a given process. This
+> restriction also applies to users that consume Protobuf files indirectly as
+> dependencies. The same extension number may be re-used across different kinds
+> of rules.
+>
+> Extension numbers may be from 1000 to 536870911, inclusive. Values from 1000
+> to 49999 are reserved for [Protobuf Global Extension Registry][1] entries, and
+> values from 50000 to 536870911 are reserved for randomly-generated integers.
```

Comment:
This isn't actually true. 50000-99999 are reserved for internal use for organizations. The random number thing is actually a somewhat-uncomfortable recommendation for me - I get the birthday paradox, but I'm not thrilled that this is what we're recommending people do. I would prefer saying "choose a number that won't conflict" and leave this up to the user.

Reply:
Thanks, I was mistaken on the ranges somehow, that's what I get for not actually checking.

I agree we shouldn't recommend it, so I've adjusted the documentation altogether:

- Fixed the ranges.
- No longer explicitly recommends randomly-generated integers
- Discourages use of 100000....536870911 for publicly-consumed schemas _at all_, due to risk of conflicts
- **But** keeping a suggestion from Miguel, I kept a note that suggests using a high-quality random source _if_ the user decides to use a randomly generated integer anyway, under the belief that it's better if they do that versus merely choose arbitrary numbers.

WDYT?

Reply:
- Remove random number generation references
- Coalesce extension ranges

Reply:
Done.